File: backend/src/main/java/com/echonovel/service/UserService.java
package com.echonovel.service;

import com.echonovel.dto.request.UserUpdateRequest;
import com.echonovel.dto.request.VipUpdateRequest;
import com.echonovel.dto.response.UserResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for user management operations.
 */
public interface UserService {

    /**
     * Update user (Role, VIP status).
     */
    UserResponse updateUser(Long userId, UserUpdateRequest request);

    /**
     * Update user VIP status (Admin).
     */
    UserResponse updateUserVipStatus(Long userId, VipUpdateRequest request);

    /**
     * Get all users (Admin).
     */
    List<UserResponse> getAllUsers();

    /**
     * Delete user (Admin).
     */
    void deleteUser(Long userId);

    /**
     * Get current user profile.
     */
    UserResponse getCurrentUserProfile(String email);

    /**
     * Update current user profile.
     */
    UserResponse updateProfile(String email, com.echonovel.dto.request.UserProfileUpdateRequest request);

    /**
     * Upload and update user avatar via Cloudinary.
     */
    UserResponse updateAvatar(String email, MultipartFile file);

    /**
     * Send OTP for change password.
     */
    void sendChangePasswordOtp(String email);

    /**
     * Change password with OTP.
     */
    void changePassword(String email, com.echonovel.dto.request.ChangePasswordRequest request);
}

File: frontend/src/services/userService.ts
import api from './api';
import type { ApiResponse, UserResponse } from '../types';

export const userService = {
  getUsers: async (): Promise<ApiResponse<UserResponse[]>> => {
    const response = await api.get<ApiResponse<UserResponse[]>>('/admin/users');
    return response.data;
  },
  
  toggleVip: async (userId: number, isVip: boolean): Promise<ApiResponse<UserResponse>> => {
    const response = await api.put<ApiResponse<UserResponse>>(`/admin/users/${userId}/vip`, { isVip });
    return response.data;
  },

  updateUser: async (userId: number, data: { role: string; vipType: string; vipExpireAt?: string }): Promise<ApiResponse<UserResponse>> => {
    const response = await api.put<ApiResponse<UserResponse>>(`/admin/users/${userId}`, data);
    return response.data;
  },

  deleteUser: async (userId: number): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/admin/users/${userId}`);
    return response.data;
  },

  getProfile: async (): Promise<ApiResponse<UserResponse>> => {
    const response = await api.get<ApiResponse<UserResponse>>('/users/profile');
    return response.data;
  },

  updateProfile: async (data: import('../types').UserProfileUpdateRequest): Promise<ApiResponse<UserResponse>> => {
    const response = await api.put<ApiResponse<UserResponse>>('/users/profile', data);
    return response.data;
  },

  sendChangePasswordOtp: async (): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/users/change-password/send-otp');
    return response.data;
  },

  changePassword: async (data: import('../types').ChangePasswordRequest): Promise<ApiResponse<void>> => {
    const response = await api.post<ApiResponse<void>>('/users/change-password', data);
    return response.data;
  },
};
